driver/chain_syncer/calldata/syncer.go

```diff
@@ -223,6 +230,136 @@ func (s *Syncer) onBlockProposed(
 	return nil
 }
 
+// handleReorg detects reorg and rewinds the chain by 1 until we find a block that is still in the chain,
+// then inserts that chain as the new head.
+func (s *Syncer) handleReorg(ctx context.Context, event *bindings.TaikoL1ClientBlockProposed) error {
+	log.Info(
+		"Reorg detected",
+		"L1Height", event.Raw.BlockNumber,
+		"L1Hash", event.Raw.BlockHash,
+		"BlockID", event.Id,
+		"Removed", event.Raw.Removed,
+	)
+
+	// rewind chain by 1 until we find a block that is still in the chain
+	var lastKnownGoodBlockId *big.Int
+	var blockId *big.Int = s.lastInsertedBlockID
+	var block *types.Block
+	var err error
+
+	for {
+		if blockId.Cmp(big.NewInt(0)) == 0 {
+			lastKnownGoodBlockId = new(big.Int).SetUint64(0)
+			break
+		}
+
+		block, err = s.rpc.L2.BlockByNumber(ctx, blockId)
+		if err != nil && !errors.Is(err, ethereum.NotFound) {
+			return err
+		}
+
+		if block != nil {
+			// block exists, we can rewind to this block
+			lastKnownGoodBlockId = blockId
+			break
+		} else {
+			// otherwise, sub 1 from blockId and try again
+			blockId = new(big.Int).Sub(s.lastInsertedBlockID, big.NewInt(1))
+		}
+	}
+
+	// shouuldnt be able to reach this error because of the 0 check above
+	// but just in case
+	if lastKnownGoodBlockId == nil {
+		return fmt.Errorf("failed to find last known good block ID after reorg")
+	}
+
+	log.Info(
+		"🔗 Last known good block ID before reorg found",
+		"blockID", event.Id,
+	)
+
+	parent, err := s.rpc.L2ParentByBlockId(ctx, lastKnownGoodBlockId)
+	if err != nil {
+		return fmt.Errorf("error getting l2 parent by block id: %w", err)
+	}
+
+	// reset l1 current to when the last known good block was inserted, and return the event.
+	blockProposedEvent, _, err := s.state.ResetL1Current(ctx, &state.HeightOrID{Height: block.Number()})
+	if err != nil {
+		return fmt.Errorf("faile to reset l1 current: %w", err)
+	}
+
+	tx, err := s.rpc.L1.TransactionInBlock(
```

Comment:
Not sure if this will work if we need to reorg to genesis? (just a corner case..)